src/storage/src/render/upsert/types.rs

```diff
@@ -7,85 +7,478 @@
+//! Implementations are required to chunk large iterators if they need to operate over smaller
+//! batches.
+//!
+//! ## `merge_snapshot_chunk`
+//! The most complicated method, this method requires implementations to consolidate a _chunk_ of
+//! updates into their state. This method effectively asks implementations to implement the logic in
+//! <https://docs.rs/differential-dataflow/latest/differential_dataflow/consolidation/fn.consolidate.html>,
+//! but under the assumption that the set of updates is a valid upsert `Collection`. Note that this
+//! allows implementations to do this a memory-efficient (or even, _memory-bounded) way. Because
+//! this is non-trivial, this module provides `StateValue`, which implements some of the core logic
+//! required to do this. `StateValue::merge_update` has more information about this.
+//!
+//! `merge_snapshot_chunk` has to return stats about the number of values and size of the state,
+//! just like `multi_put`.
+//!
+//! Another curiosity is that implementation can assume that `merge_snapshot_chunk` is called with
+//! a set of updates with a number of keys not greater than `UpsertState::SNAPSHOT_BATCH_SIZE`. This
+//! is different than `multi_put` and `multi_get` purely because it simplifies the way that the `upsert`
+//! operator handles snapshots.
+//!
+//!
+//! # A note on state size
+//!
+//! The `UpsertState` trait requires implementations report _relatively accurate_ information about
+//! how the state size changes over time. Note that it does NOT ask the implementations to give
+//! accurate information about actual resource consumption (like disk space including space
+//! amplification), and instead is just asking about the size of the values, after they have been
+//! encoded. For implementations like `RocksDB`, these may be highly accurate (it literally
+//! reports the encoded size as written to the RocksDB API, and for others like the
+//! `InMemoryHashMap`, they may be rough estimates of actual memory usage. See
+//! `StateValue::memory_size` for more information.
+//!
+//! Note also that after snapshot consolidation, additional space may be used if `StateValue` is
+//! used.
+
+use std::collections::hash_map;
+use std::num::Wrapping;
 use std::sync::Arc;
 use std::time::Instant;
 
+use bincode::Options;
 use itertools::Itertools;
 
-use mz_ore::cast::CastLossy;
+use mz_ore::cast::{CastFrom, CastLossy};
 use mz_ore::collections::HashMap;
 
 use super::{UpsertKey, UpsertValue};
 use crate::source::metrics::UpsertSharedMetrics;
 
+/// The default set of `bincode` options used for consolidating
+/// upsert snapshots (and writing values to RocksDB).
+pub type BincodeOpts = bincode::config::WithOtherTrailing<
+    bincode::config::DefaultOptions,
+    bincode::config::AllowTrailing,
+>;
+
+/// Build the default `BincodeOpts`.
+pub fn upsert_bincode_opts() -> BincodeOpts {
+    bincode::DefaultOptions::new().allow_trailing_bytes()
```

Comment:
We no longer need this because we track the length of the XOR buffer so we can just the normal bincode options. In my original prototype I didn't track the length which is why I had to do this